### PR TITLE
[FEAT] Optimization improvements and new features

### PR DESCRIPTION
**Pull request by cchallu**
- Added new optimization features and improvements to all models:
  * learning rate scheduler: added scheduler and new hyperparameter to control the number of decays.
  * early stropping: early stopping using validation loss, new patience hyperparameter.
  * val_check_steps: new hyperparameter to control how often to check validation loss.
 
- Training is determined by steps, including the scheduler, early stopping, and validation loss.
  * Added explicit max_steps hyperparameter to all models.
  * Added warning max_epochs hyperparameter, which will be deprecated.
  * Modified tests, default search spaces, and common base classes to reflect new features.
 
 - Removed cv-performance CI test.

### CODE REVIEW

File: action_files/test_models/src/models.py [hunk 1]
```diff
@@ -35,32 +35,35 @@ def main(dataset: str = 'M3', group: str = 'Other') -> None:
...
+                   'max_steps': 50,
+                   "val_check_steps": 10}
     models = [
-        DilatedRNN(h=horizon, input_size=2 * horizon, encoder_hidden_size=50, max_epochs=50),
-        RNN(h=horizon, input_size=2 * horizon, encoder_hidden_size=50, max_epochs=50),
-        TCN(h=horizon, input_size=2 * horizon, encoder_hidden_size=20, max_epochs=100),
-        LSTM(h=horizon, input_size=2 * horizon, encoder_hidden_size=50, max_epochs=50),
-        GRU(h=horizon, input_size=2 * horizon, encoder_hidden_size=50, max_epochs=50),
+        DilatedRNN(h=horizon, input_size=2 * horizon, encoder_hidden_size=50, max_steps=50),
+        RNN(h=horizon, input_size=2 * horizon, encoder_hidden_size=50, max_steps=50),
```
Comment (kdgutier) [inline action_files/test_models/src/models.py:54]: Is it the case that  max_epochs > max_steps?
Did we confirm that CI evaluation maintains reasonable performance?

File: nbs/models.dilated_rnn.ipynb [hunk 2]
```diff
@@ -375,7 +375,11 @@
     "    `hist_exog_list`: str list, historic exogenous columns.<br>\n",
     "    `stat_exog_list`: str list, static exogenous columns.<br>\n",
     "    `loss`: PyTorch module, instantiated train loss class from [losses collection](https://nixtla.github.io/neuralforecast/losses.pytorch.html).<br>\n",
-    "    `learning_rate`: float=1e-3, initial optimization learning rate (0,1).<br>\n",
+    "    `max_steps`: int, maximum number of training steps.<br>\n",
+    "    `learning_rate`: float, Learning rate between (0, 1).<br>\n",
```
Comment (kdgutier) [inline nbs/models.dilated_rnn.ipynb:379]: Can we add the default values to the docstrings?

File: nbs/models.gru.ipynb [hunk 3]
```diff
@@ -108,7 +108,11 @@
     "    `hist_exog_list`: str list, historic exogenous columns.<br>\n",
     "    `stat_exog_list`: str list, static exogenous columns.<br>\n",
     "    `loss`: PyTorch module, instantiated train loss class from [losses collection](https://nixtla.github.io/neuralforecast/losses.pytorch.html).<br>\n",
-    "    `learning_rate`: float=1e-3, initial optimization learning rate (0,1).<br>\n",
+    "    `max_steps`: int, maximum number of training steps.<br>\n",
+    "    `learning_rate`: float, Learning rate between (0, 1).<br>\n",
```
Comment (kdgutier) [inline nbs/models.gru.ipynb:112]: Similar comment to that of `dilated_rnn.ipynb`.
Can we add the values of the defaults in the docstring?

File: neuralforecast/common/_base_recurrent.py [hunk 4]
```diff
@@ -48,6 +54,16 @@ def __init__(
...
         self.loss = loss
         self.learning_rate = learning_rate
+        self.max_steps = max_steps
+        self.num_lr_decays = num_lr_decays
+        self.lr_decay_steps = (
+            max(max_steps // self.num_lr_decays, 1)
+            if self.num_lr_decays > 0
+            else max_steps
+        )
+        self.early_stop_patience_steps = early_stop_patience_steps
```
Comment (kdgutier) [inline neuralforecast/common/_base_recurrent.py:64]: What happens with `EarlyStopping(patience=-1)`
Should we overwrite `early_stop_patience_steps` when it is negative?

File: neuralforecast/common/_base_recurrent.py [hunk 5]
```diff
@@ -96,7 +123,15 @@ def on_fit_start(self):
         random.seed(self.random_seed)
 
     def configure_optimizers(self):
-        return torch.optim.Adam(self.parameters(), lr=self.learning_rate)
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.learning_rate)
+        scheduler = {
+            "scheduler": torch.optim.lr_scheduler.StepLR(
+                optimizer, self.lr_decay_steps, 0.5
```
Comment (kdgutier) [inline neuralforecast/common/_base_recurrent.py:129]: to have better code readability and protect us agains inputs ordering changes
can we add the name of the arguments?

https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html

File: action_files/test_models/src/models.py (hunk 1 above)
Comment (cchallu) [reply action_files/test_models/src/models.py]: Done

File: nbs/models.dilated_rnn.ipynb (hunk 2 above)
Comment (cchallu) [reply nbs/models.dilated_rnn.ipynb:379]: Done

File: nbs/models.gru.ipynb (hunk 3 above)
Comment (cchallu) [reply nbs/models.gru.ipynb]: Done

File: neuralforecast/common/_base_windows.py [hunk 6]
```diff
@@ -50,6 +56,15 @@ def __init__(
...
         self.loss = loss
         self.learning_rate = learning_rate
+        self.max_steps = max_steps
+        self.num_lr_decays = num_lr_decays
+        self.lr_decay_steps = (
+            max(max_steps // self.num_lr_decays, 1)
+            if self.num_lr_decays > 0
+            else max_steps
+        )
+        self.early_stop_patience_steps = early_stop_patience_steps
```
Comment (kdgutier) [inline neuralforecast/common/_base_windows.py:66]: Similar comment to that of _base_recurrent
What happens with default negative?

File: neuralforecast/common/_base_windows.py [hunk 7]
```diff
@@ -103,7 +130,15 @@ def on_fit_start(self):
         random.seed(self.random_seed)
 
     def configure_optimizers(self):
-        return torch.optim.Adam(self.parameters(), lr=self.learning_rate)
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.learning_rate)
+        scheduler = {
+            "scheduler": torch.optim.lr_scheduler.StepLR(
+                optimizer, self.lr_decay_steps, 0.5
```
Comment (kdgutier) [inline neuralforecast/common/_base_windows.py:136]: to have better code readability and protect us agains inputs ordering changes
can we add the name of the arguments?

https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html

**Review by kdgutier — APPROVED** [review]

File: neuralforecast/common/_base_recurrent.py (hunk 4 above)
Comment (cchallu) [reply neuralforecast/common/_base_recurrent.py:64]: Nothing, see if below in definition of callbacks

File: neuralforecast/common/_base_windows.py (hunk 7 above)
Comment (cchallu) [reply neuralforecast/common/_base_windows.py:136]: Done

File: neuralforecast/common/_base_recurrent.py (hunk 5 above)
Comment (cchallu) [reply neuralforecast/common/_base_recurrent.py:129]: Done
